### DISCUSSION TITLE
Is is possible to launch AR mode automatically?

### DISCUSSION THREAD

**Discussion by hbaugmentedreality** [google/model-viewer#2514] [Q&A]
I mean, when I load the webpage I would like to go to AR mode automatically, without pressing any "view in your space" button.  Am I clear?  Is is possible? 
Thanks!!!

**Comment by elalish** [comment]
No, it's not allowed by Chrome and other browser's security and privacy policies.

**Reply by david-rhodes** [reply]
I believe this generally works for scene-viewer and QuickLook, but not WebXR, right? I mean, that may change in the future, though...

**Reply by elalish** [reply]
I don't think you can intent out to an app like Scene Viewer without user activation either, right? With Quick Look, I'm not sure what Safari allows.

**Reply by david-rhodes** [reply]
Yes, you can. On window load, I simply call `modelViewer.activateAR()` and it works as expected so long as I'm not using WebXR. QuickLook is a bit trickier because `canActivateAR` returns `false` even once the page is loaded. So you have to poll that property or otherwise wait until the capability has been correctly determined (which is why I have previously asked when it's appropriate to query `canActivateAR`).

**Reply by elalish** [reply]
Interesting, thanks!

**Reply by hbaugmentedreality** [reply]
Thanks @david-rhodes .   Sorry about my basic knowledge.   Can you give me a small html example code of how to do this?  Thank you very much.

**Reply by david-rhodes** [reply]
No problem, @hbaugmentedreality! Just use the following HTML script:

```html
<script>
  window.addEventListener("load", () => {
    var modelViewer = document.querySelector('model-viewer');
    console.log("Can activate AR: " + modelViewer.canActivateAR);
    modelViewer.activateAR();
  });
</script>
```

And ensure that your `ar-modes="scene-viewer quick-look"` (no WebXR).

For iOS, you can probably use [something like this](https://stackoverflow.com/a/8682723) to poll `modelViewer.canActivateAR` to know when to execute `modelViewer.activateAR()`. This is because (for me), `canActivateAR` returns `false` when the window is initially loaded. There's some strange async behaviour that causes this.

**Reply by Sov3rain** [reply]
Hey @david-rhodes, I tried your snipet on multiple Android devices but I can't make it work. Trying to autolaunch the AR view results on the phone falling back to 3D view instead with no error message in console.

When I enable WebXR fallback, it tells me that in order to launch an AR session, I need the user to interact with the webpage. Does browsers now blocks this kind of "autolaunch" tricks?

EDIT: on iOS safari it works fine. But no luck on any Android phone in Chrome that I tried.

**Reply by david-rhodes** [reply]
Unfortunately, maybe it was never intended to work that way on Android. It could have been patched. As mentioned, it has never worked with WebXR. 

Check your logs to see what `canActivateAR` is returning when it fails.

**Reply by Sov3rain** [reply]
When removing the WebXR option, I got nothing, I'm not redirected to AR mode when hitting the web page, nor do I have the integrated AR button on the viewer (testing on a Samsung Galaxy S20), even if the phone is clearly compatible with scene-viewer.

When adding WebXR option, I got both button and redirection, but I'm redirected to the fallback URL (AR Core doc page). I know that it's not intended to work, but that's all I've got so far.

`canActivateAR` always return true though. I'm a bit lost here.

**Reply by harowosegbe** [reply]
@elalish , checking to see if we can have an attribute `ar-auto-launch`, if this is present we can use the default browser dialog to request if the user wants to go into AR, and we can trigger AR from here with the user interaction. This approach should work for both native and webxr.

**Reply by elalish** [reply]
PRs welcome! I've never gotten that to work, but maybe you know something I don't.

**Reply by harowosegbe** [reply]
@elalish sweet, I will look into making a PR
